src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs

```diff
@@ -592,4 +599,170 @@ private void FreeAsmMeta(IntPtr asmMetaPtr)
+    {
+        class HeaderInfo
+        {
+            public uint VirtualAddress;
+            public uint Size;
+            public uint FileOffset;
+        }
+
+        /// <summary>
+        /// Given a path get the CLR runtime version of the file
+        /// </summary>
+        /// <param name="path">path to the file</param>
+        /// <returns>The CLR runtime version or empty if the path does not exist or the file is not an assembly.</returns>
+        public static string GetRuntimeVersion(string path)
+        {
+            using (var sr = new BinaryReader(File.OpenRead(path)))
+            {
+                if (!File.Exists(path))
+                    return string.Empty;
+                
+                try
+                {
+                    const uint PEHeaderPointerOffset = 0x3c;
+                    const uint PEHeaderSize = 20;
+                    const uint OptionalPEHeaderSize = 224;
+                    const uint OptionalPEPlusHeaderSize = 240;
+                    const uint SectionHeaderSize = 40;
+
+                    // There must be room for at least the PE header offset and size, and sections
+                    if (sr.BaseStream.Length < PEHeaderPointerOffset + 4 + PEHeaderSize + OptionalPEHeaderSize + SectionHeaderSize)
+                        return string.Empty;
+
+                    // PE header
+                    sr.BaseStream.Position = PEHeaderPointerOffset;
+                    var peHeaderOffset = sr.ReadUInt32();
+
+                    if (peHeaderOffset + 4 + PEHeaderSize + OptionalPEHeaderSize + SectionHeaderSize >= sr.BaseStream.Length)
+                        return string.Empty;
+
+                    sr.BaseStream.Position = peHeaderOffset;
+                    if (!ReadBytes(sr, (byte)'P', (byte)'E', 0, 0)) // PE header signature
+                        return string.Empty;
+
+                    peHeaderOffset += 4; // Skip signature
+
+                    // Read the number of sections
+                    sr.BaseStream.Position = peHeaderOffset + 2;
+                    var numberOfSections = sr.ReadUInt16();
+                    if (numberOfSections > 96)
+                        return string.Empty; // There can't be more than 96 sections, something is wrong
+
+                    // Optional header start
+                    var optionalHeaderOffset = peHeaderOffset + PEHeaderSize;
+
+                    uint cliHeaderRvaOffset;
+                    uint optionalPEHeaderSize;
+
+                    // Get the offset to the CLI header RVA. That's in the optional header,
+                    // which can have two formats, PE32 or PE32+
+                    sr.BaseStream.Position = optionalHeaderOffset;
+                    var magicNumber = sr.ReadUInt16();
+
+                    if (magicNumber == 0x10b) // PE32
+                    {
+                        optionalPEHeaderSize = OptionalPEHeaderSize;
+                        cliHeaderRvaOffset = optionalHeaderOffset + 208;
```

Comment:
I think this makes the code more readable since it is easier to compare the offset to what's in the specs. I added more comments to make easier to understand what those offsets are.